elastic/kibana
# [Response Ops][Task Manager] Not showing unrecognized task types in the health summary

## Summary

With the [PR](https://github.com/elastic/kibana/pull/187999) to add resource based claiming to task manager, we updated the health summary to include the cost of tasks. This caused issues in environments where certain tasks might exist but their task definitions are no longer registered due to disabled plugins. The ES aggregation just looks at the tasks that are in the TM index and then tries to grab their cost from their task definition, which does not exist, causing a `WorkloadAggregator` error. These errors are logged on an hourly interval for background task only nodes so we see an uptick of errors logged in serverless. This does not affect actual task manager operations.

This PR adds a try/catch around getting the task definition and omits unregistered task types from the health summary. I will open up a separate issue to discuss whether we should remove throwing an error for unregistered types in favor of handling it more gracefully.

I also checked and we have a functional test for removed and unrecognized task types here: https://github.com/elastic/kibana/blob/main/x-pack/test/plugin_api_integration/test_suites/task_manager/task_management_removed_types.ts

## To Verify
I was able to see the `WorkloadAggregator` error when running the functional test by making a few changes:
- remove the check to only log for background task nodes

```
--- a/x-pack/plugins/task_manager/server/plugin.ts
+++ b/x-pack/plugins/task_manager/server/plugin.ts
@@ -236,6 +236,7 @@ export class TaskManagerPlugin
     if (this.isNodeBackgroundTasksOnly()) {
       setupIntervalLogging(monitoredHealth$, this.logger, LogHealthForBackgroundTasksOnlyMinutes);
     }
+    setupIntervalLogging(monitoredHealth$, this.logger, LogHealthForBackgroundTasksOnlyMinutes);
```

- reduce the logging interval

```
--- a/x-pack/plugins/task_manager/server/lib/log_health_metrics.ts
+++ b/x-pack/plugins/task_manager/server/lib/log_health_metrics.ts
@@ -35,7 +35,8 @@ export function setupIntervalLogging(
     monitoredHealth = m;
   });

-  setInterval(onInterval, 1000 * 60 * minutes);
+  // setInterval(onInterval, 1000 * 60 * minutes);
+  setInterval(onInterval, 1000 * 15);

   function onInterval() {
```

- remove the `after` block that removes the removed type from after the test

```
--- a/x-pack/test/plugin_api_integration/test_suites/task_manager/task_management_removed_types.ts
+++ b/x-pack/test/plugin_api_integration/test_suites/task_manager/task_management_removed_types.ts
@@ -52,9 +52,9 @@ export default function ({ getService }: FtrProviderContext) {
       await esArchiver.load('x-pack/test/functional/es_archives/task_manager_removed_types');
     });

-    after(async () => {
-      await esArchiver.unload('x-pack/test/functional/es_archives/task_manager_removed_types');
-    });
+    // after(async () => {
+    //   await esArchiver.unload('x-pack/test/functional/es_archives/task_manager_removed_types');
+    // });

     function scheduleTask(
```

You should see the `WorkloadAggregator` error show up in the test logs when on `main` but not in this PR.
